### PR TITLE
fix: crash if samples have different amount of costs associated

### PR DESCRIPTION
**Pull request by lievenhey**
If we have a recording with the following layout: A B C where A: part where function f is accessed
B: list samples
C: part where function g is accessed
then f won't have the lost events cost while g does. This will cause a crash when the model tries to access it

fixes: #629

### CODE REVIEW

File: src/models/callercalleemodel.h [hunk 1]
```diff
@@ -263,16 +263,26 @@ class LocationCostModelImpl : public HashModel<Data::SourceLocationCostMap, Mode
     QVariant cell(int column, int role, const Data::FileLine& fileLine,
                   const Data::LocationCost& costs) const final override
     {
+        // if we have a recording with the following layout: A B C where
+        // A: part where function f is accessed
+        // B: list samples
```
Comment (milianw) [inline src/models/callercalleemodel.h:268]: list -> lost

also in the commit message

File: src/models/callercalleemodel.h [hunk 2]
```diff
@@ -263,16 +263,26 @@ class LocationCostModelImpl : public HashModel<Data::SourceLocationCostMap, Mode
...
+        // C: part where function g is accessed
+        // then f won't have the lost events cost while g does.
+        // This will cause a crash when the model tries to access it
+
+        const std::size_t costType = column - NUM_BASE_COLUMNS;
```
Comment (milianw) [inline src/models/callercalleemodel.h:273]: const auto

File: src/models/callercalleemodel.h [hunk 3]
```diff
@@ -263,16 +263,26 @@ class LocationCostModelImpl : public HashModel<Data::SourceLocationCostMap, Mode
...
+        auto safeSelfCost = costType < costs.selfCost.size() ? costs.selfCost[costType] : 0;
+        auto safeInclusiceCost = costType < costs.inclusiveCost.size() ? costs.selfCost[costType] : 0;
```
Comment (milianw) [inline src/models/callercalleemodel.h:275]: typo: inclusive

File: src/models/callercalleemodel.h [hunk 4]
```diff
@@ -263,16 +263,26 @@ class LocationCostModelImpl : public HashModel<Data::SourceLocationCostMap, Mode
...
     {
+        // if we have a recording with the following layout: A B C where
+        // A: part where function f is accessed
+        // B: list samples
+        // C: part where function g is accessed
+        // then f won't have the lost events cost while g does.
+        // This will cause a crash when the model tries to access it
+
+        const std::size_t costType = column - NUM_BASE_COLUMNS;
+        auto safeSelfCost = costType < costs.selfCost.size() ? costs.selfCost[costType] : 0;
```
Comment (milianw) [inline src/models/callercalleemodel.h:274]: there are many more places where we access costs in such a way, and all of them would need to be updated, no? 

please make `ItemCost` a real wrapper class around `valarray` instead of just a typedef, then make `operator[]` fail safe

**Review by milianw — CHANGES_REQUESTED** [review]

File: src/models/callercalleemodel.h (hunk 2 above)
Comment (lievenhey) [reply src/models/callercalleemodel.h:273]: `std::valarray<>::opetator[]` takes `std::size_t` as argument so I need this to be `std::size_t` otherwise `clang-tidy` complains (auto will set the type to `int`)

File: src/models/data.h [hunk 5]
```diff
@@ -203,7 +203,32 @@ struct FrameLocation
...
 };
 
-using ItemCost = std::valarray<qint64>;
+class ItemCost
+{
+public:
+    ItemCost(std::size_t size = 0);
+    ItemCost(qint64 value, std::size_t count);
+    ItemCost(std::valarray<qint64>&& cost);
+    ~ItemCost();
```
Comment (milianw) [inline src/models/data.h:212]: remove, rule of zero

File: src/models/data.cpp [hunk 6]
```diff
@@ -12,6 +12,79 @@
 
 using namespace Data;
 
+ItemCost::ItemCost(std::size_t size)
```
Comment (milianw) [inline src/models/data.cpp:15]: inline all of this

File: src/models/data.cpp [hunk 7]
```diff
@@ -12,6 +12,79 @@
...
+}
+
+std::size_t ItemCost::size() const
+{
+    return m_cost.size();
+}
+
+ItemCost ItemCost::operator+(const ItemCost& rhs) const
+{
+    return static_cast<std::valarray<qint64>>(m_cost + rhs.m_cost);
```
Comment (milianw) [inline src/models/data.cpp:43]: out of interest: why the static cast here and below?

Comment (lievenhey) [reply src/models/data.cpp:43]: because the resulting type is `std::_Expr<std::__detail::_BinClos<std::__minus, std::_ValArray, std::_ValArray, long long int, long long int>, long long int>`

Comment (lievenhey) [reply src/models/data.cpp:43]: so I am casting that to `std::valarray` and then the constructor will turn that into `ItemCost`

Comment (milianw) [reply src/models/data.cpp:27]: this should work too, no?

```
return ItemCost(m_cost + rhs.m_cost);
```

File: src/models/data.cpp [hunk 8]
```diff
@@ -338,28 +395,21 @@ void Data::callerCalleesFromBottomUpData(const BottomUpResults& bottomUpData, Ca
 
 QDebug Data::operator<<(QDebug stream, const Symbol& symbol)
 {
-    stream.noquote().nospace() << "Symbol{"
-                               << "symbol=" << symbol.symbol << ", "
-                               << "relAddr=" << symbol.relAddr << ", "
-                               << "size=" << symbol.size << ", "
-                               << "binary=" << symbol.binary << "}";
+    stream.noquote().nospace() << "Symbol{" << "symbol=" << symbol.symbol << ", " << "relAddr=" << symbol.relAddr
```
Comment (milianw) [inline src/models/data.cpp]: unrelated changes

File: src/models/data.h [hunk 9]
```diff
@@ -203,7 +203,42 @@ struct FrameLocation
...
+        : m_cost(size)
+    {
+    }
+
+    ItemCost(qint64 value, std::size_t count)
+        : m_cost(value, count)
+    {
+    }
+
+    ItemCost(std::valarray<qint64>&& cost)
```
Comment (milianw) [inline src/models/data.h:219]: take by value

File: src/models/data.h [hunk 10]
```diff
@@ -203,7 +203,42 @@ struct FrameLocation
...
+        : m_cost(cost)
```
Comment (milianw) [inline src/models/data.h:220]: std::move here

File: src/models/data.h [hunk 11]
```diff
@@ -203,7 +203,42 @@ struct FrameLocation
...
+    {
+    }
+
+    void resize(std::size_t newSize, qint64 value = 0);
+    std::size_t size() const;
+
+    ItemCost operator+(const ItemCost& rhs) const;
+    ItemCost operator-(const ItemCost& rhs) const;
+    ItemCost& operator+=(const ItemCost& rhs);
```
Comment (milianw) [inline src/models/data.h:229]: missing -= for symmetry

File: src/models/data.cpp [hunk 12]
```diff
@@ -12,6 +12,63 @@
 
 using namespace Data;
 
+void ItemCost::resize(std::size_t newSize, qint64 value)
```
Comment (milianw) [inline src/models/data.cpp:15]: as I said before: all of these should be inline

**Review by milianw — CHANGES_REQUESTED** [review]

File: src/models/data.cpp (hunk 12 above)
Comment (lievenhey) [reply src/models/data.cpp:15]: ah, that wasn't clear, will do

**Review by milianw — APPROVED** [review]
lgtm, thanks
